contracts/farm/LiquidityMining.sol

```diff
@@ -0,0 +1,607 @@
+	 * @param _user the address of user, tokens will be deposited to it or to msg.sender
+	 * @param alreadyTransferred true if the pool tokens have already been transferred
+	 */
+	function _deposit(address _poolToken, uint256 _amount, address _user, bool alreadyTransferred) internal{
+		require(poolIdList[_poolToken] != 0, "Pool token not found");
+		address userAddress = _user != address(0) ? _user : msg.sender;
+
+		uint256 poolId = _getPoolId(_poolToken);
+		PoolInfo storage pool = poolInfoList[poolId];
+		UserInfo storage user = userInfoMap[poolId][userAddress];
+
+		_updatePool(poolId);
+		//sends reward directly to the user
+		_updateReward(pool, user);
+
+		if (_amount > 0) {
+			//receives pool tokens from msg.sender, it can be user or WrapperProxy contract
+			if(!alreadyTransferred)
+				pool.poolToken.safeTransferFrom(address(msg.sender), address(this), _amount);
+			user.amount = user.amount.add(_amount);
+		}
+		_updateRewardDebt(pool, user);
+		emit Deposit(userAddress, _poolToken, _amount);
+	}
+
+	/**
+	 * @notice transfers reward tokens
+	 * @param _poolToken the address of pool token
+	 * @param _user the address of user to claim reward from (can be passed only by wrapper contract)
+	 */
+	function claimReward(address _poolToken, address _user) public {
+		require(poolIdList[_poolToken] != 0, "Pool token not found");
+		address userAddress = _getUserAddress(_user);
+
+		uint256 poolId = _getPoolId(_poolToken);
+		PoolInfo storage pool = poolInfoList[poolId];
+		UserInfo storage user = userInfoMap[poolId][userAddress];
+
+		_updatePool(poolId);
+		_updateReward(pool, user);
+		_transferReward(user, userAddress, true);
+		_updateRewardDebt(pool, user);
+	}
+
+	/**
+	 * @notice withdraws pool tokens and transfers reward tokens
+	 * @param _poolToken the address of pool token
+	 * @param _amount the amount of pool tokens
+	 * @param _user the user address will be used to process a withdrawal (can be passed only by wrapper contract)
+	 */
+	function withdraw(
+		address _poolToken,
+		uint256 _amount,
+		address _user
+	) public {
+		require(poolIdList[_poolToken] != 0, "Pool token not found");
+		address userAddress = _getUserAddress(_user);
+
+		uint256 poolId = _getPoolId(_poolToken);
+		PoolInfo storage pool = poolInfoList[poolId];
+		UserInfo storage user = userInfoMap[poolId][userAddress];
+		require(user.amount >= _amount, "Not enough balance");
+
+		_updatePool(poolId);
+		_updateReward(pool, user);
+		_transferReward(user, userAddress, false);
+
+		user.amount = user.amount.sub(_amount);
+		pool.poolToken.safeTransfer(address(msg.sender), _amount); //sent to the user or wrapper
+		_updateRewardDebt(pool, user);
+		emit Withdraw(userAddress, _poolToken, _amount);
+	}
+
+	function _getUserAddress(address _user) internal view returns (address) {
+		address userAddress = msg.sender;
+		if (_user != address(0)) {
+			//only wrapper can pass _user parameter
+			require(msg.sender == wrapper && _user == tx.origin, "unauthorized");
```

Comment:
`tx.origin` is often not the owner of tokens, even without complex DeFi integrations:
- there are relayers and metatransactions
- there are smart contract wallets, multisigs etc

requiring the user to `tx.origin` will break some usecases and will probably lock user's tokens: `_getUserAddress` is called to `claimReward` and `withdraw`, but it is not required to deposit tokens.

Reply:
Removed check of tx.origin